File: leetcode/0804. Unique Morse Code Words/Solution.swift
// https://leetcode.com/problems/unique-morse-code-words/

class Solution {
    let m = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]

    func tranformation(_ word: String) -> String {
        var r = ""
        for c in word {
            let diff = Int((c.asciiValue ?? 0) - (Character("a").asciiValue ?? 0))
            r += m[diff]
        }

        return r
    }

    func uniqueMorseRepresentations(_ words: [String]) -> Int {
        var s = Set<String>()
        for word in words {
            s.insert(tranformation(word))
        }

        return s.count
    }
}

let s = Solution()
assert(s.uniqueMorseRepresentations(["gin", "zen", "gig", "msg"]) == 2)

File: leetcode/0804. Unique Morse Code Words/Solution.py
# https://leetcode.com/problems/unique-morse-code-words/

from typing import List

class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        m = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]

        s = set()
        for word in words:
            morse_c_list = [m[ord(c) - ord("a")] for c in word]
            word = "".join(morse_c_list)
            s.add(word)

        return len(s)

s = Solution()
assert s.uniqueMorseRepresentations(["gin", "zen", "gig", "msg"]) == 2